# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [1]:
lambda arguments: expression


<function __main__.<lambda>(arguments)>

#### Saving to a Variable

In [2]:

is_even = lambda x: x % 2 == 0


numbers = [1, 2, 3, 4, 5, 6]
even_numbers = list(filter(is_even, numbers))
print(even_numbers)  


[2, 4, 6]


#### Multiple Inputs

In [3]:

add = lambda x, y: x + y

result = add(2, 3)
print(result)  


5


#### Passing a Lambda into a Function

In [21]:

def apply_function(func, values):
    return [func(value) for value in values]

squared_= apply_function(lambda x: x**2, [1, 2, 3, 4])
print(squared_) 

doubled_ = apply_function(lambda x: x * 2, [1, 2, 3, 4])
print(doubled_)  


[1, 4, 9, 16]
[2, 4, 6, 8]


#### Returning a Lambda from a Function

In [ ]:
def create_multiplier(factor):
    return lambda x: x * factor


doubler = create_multiplier(2)

tripler = create_multiplier(3)
print(tripler(5))  


#### If Statements within Lambdas

In [ ]:

is_even = lambda x: "Even" if x % 2 == 0 else "Odd"

print(is_even(4))  
print(is_even(5)) 


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [ ]:
f = lambda x: x**3

print(f(2)) 
print(f(3))  


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [5]:
map(function to be used, list to be used)


SyntaxError: invalid syntax. Perhaps you forgot a comma? (1241297438.py, line 1)

#### Using Lambda's with Map

In [6]:

numbers = [1, 2, 3, 4, 5]
cubed_numbers = map(lambda x: x**3, numbers)

cubed_numbers_list = list(cubed_numbers)
print(cubed_numbers_list)  


[1, 8, 27, 64, 125]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [7]:

numbers = [1, 2, 3, 4, 5]

modified_numbers = list(map(lambda x: 2*x - 1, numbers))

print(modified_numbers)  


[1, 3, 5, 7, 9]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [ ]:
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

filtered_numbers = list(filter(lambda x: x % 2 == 0, numbers))

print(filtered_numbers)  


#### Using Lambda's with Filter()

In [9]:
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

filtered_numbers = list(filter(lambda x: x % 2 == 0, numbers))


print(filtered_numbers) 

[2, 4, 6, 8, 10]


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [8]:
import statistics

numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

mean = statistics.mean(numbers)

filtered_numbers = list(filter(lambda x: x >= mean, numbers))
print(filtered_numbers)


[6, 7, 8, 9, 10]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [25]:
import functools

# Syntax for reduce() function
functools.reduce(function, iterable[, initializer])


SyntaxError: invalid syntax (501301428.py, line 4)

#### Using Lambda's with Reduce()

In [11]:
import functools

numbers = [1, 2, 3, 4, 5]

sum_result = functools.reduce(lambda x, y: x + y, numbers)

print(sum_result) 


15


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [12]:
import functools

numbers = [1, 2, 3, 4, 5]

product_result = functools.reduce(lambda x, y: x * y, numbers)

print(product_result)  


120


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [ ]:
def factorial(n):
    if n == 0 or n == 1:
        return 1
    
    else:
        return n * factorial(n - 1)


result = factorial(5)
print(result)  


#### Writing a Factorial Function

In [13]:
def factorial(n):
    if n == 0 or n == 1:
        return 1

    else:
        return n * factorial(n - 1)

print(factorial(5))  


120


#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [ ]:
def subtract_all(n):
    if n == 1:
        return 1

    else:
        return n - subtract_all(n - 1)

print(subtract_all(5))  


## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [ ]:
def generate_numbers(n):
    for i in range(n):
        yield i

gen = generate_numbers(5)

for number in gen:
    print(number)


#### Infinite Generator

In [20]:
def infinite_generator():
    i = 0
    while True:
        yield i
        i += 1

gen = infinite_generator()

for _ in range(10):
    print(next(gen))


0
1
2
3
4
5
6
7
8
9


#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [23]:
def squared_(n):
    while n >= 0:
        yield n * n
        n -= 1

gen = squared_(5)

for square in gen:
    print(square)


25
16
9
4
1
0


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [22]:
places = [" ", "Argentina", " ", "San Diego", "", "  ", "", "Boston", "New York"]

filtered_ = list(filter(lambda x: x.strip() != "", places))

print(filtered_)


['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [17]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield", "David hassELHOFF", "Gary A.J. Bernstein"]
author.sort(key=lambda x: x.split()[-1].lower())

print(author)


['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [16]:
location = [('Nashua', 32), ("Boston", 12), ("Los Angeles", 44), ("Miami", 29)]

converted_temperatures = list(map(lambda x: (x[0], (9/5)*x[1] + 32), location))

print(converted_temperatures)


[('Nashua', 89.6), ('Boston', 53.6), ('Los Angeles', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [15]:
def fibonacci(n):
   
    if n == 0:
        return 1
    elif n == 1:
        return 1
    else:
        return fibonacci(n - 1) + fibonacci(n - 2)
 
def print_fibonacci_sequence(n):
    for i in range(n + 1):
        print(f"Iteration {i}:{fibonacci(i)}")

print_fibonacci_sequence(5)


Iteration 0:1
Iteration 1:1
Iteration 2:2
Iteration 3:3
Iteration 4:5
Iteration 5:8
